### DISCUSSION TITLE
How to add user ?

### DISCUSSION THREAD

**Discussion by ArnabDom** [verdaccio/verdaccio#3433] [Q&A] [answered]
Please Help!!! 
Thank you in Advance :)

Currently, i was testing in my local pc/laptop
I was testing with 2 packages 1 for $all users and another package for only $authenticated users
i can not login to my previous users in web and console and
every time i try to add a new user, i am getting this message on my terminal

administrator@swift-sf314-42:~$ npm adduser --registry http://localhost:4873
npm notice Log in on http://localhost:4873/
Username: user2324
Password: 
Email: (this IS public) test@gmail.com
npm ERR! code E409
npm ERR! 409 Conflict - PUT http://localhost:4873/-/user/org.couchdb.user:user2324 - bad username/password, access denied

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/administrator/.npm/_logs/2022-10-12T10_53_13_667Z-debug-0.log

Here is the Live/ Local Server Terminal output

administrator@swift-sf314-42:~$ verdaccio
 warn --- config file  - /home/administrator/.config/verdaccio/config.yaml
 info --- plugin successfully loaded: verdaccio-audit
 warn --- http address - http://localhost:4873/ - verdaccio/5.15.4
 http --- 127.0.0.1 requested 'POST /-/v1/login'
 error--- authenticating for user arnabgg failed. Error: bad username/password, access denied
(node:139175) [DEP0106] DeprecationWarning: crypto.createDecipher is deprecated.
 http --- 404, user: null(127.0.0.1), req: 'POST /-/v1/login', bytes: 29/150
 http --- 127.0.0.1 requested 'PUT /-/user/org.couchdb.user:user2324'
 error--- authenticating for user arnabgg failed. Error: bad username/password, access denied
 warn --- the plugin method add_user in the auth plugin is deprecated and will be removed in next major release, notify to the plugin author
 error--- the user user2324 could not being added. Error: bad username/password, access denied
 http --- 409, user: null(127.0.0.1), req: 'PUT /-/user/org.couchdb.user:user2324', error: bad username/password, access denied
 http --- 127.0.0.1 requested 'GET /-/user/org.couchdb.user:user2324?write=true'
 error--- authenticating for user arnabgg failed. Error: bad username/password, access denied
 http --- 200, user: null(127.0.0.1), req: 'GET /-/user/org.couchdb.user:user2324?write=true', bytes: 0/45
 http --- 127.0.0.1 requested 'PUT /-/user/org.couchdb.user:user2324/-rev/undefined'
 error--- authenticating for user user2324 failed. Error: bad username/password, access denied
 warn --- the plugin method add_user in the auth plugin is deprecated and will be removed in next major release, notify to the plugin author
 error--- the user user2324 could not being added. Error: bad username/password, access denied
 http --- 409, user: null(127.0.0.1), req: 'PUT /-/user/org.couchdb.user:user2324/-rev/undefined', error: bad username/password, access denied
 http --- 127.0.0.1 requested 'PUT /-/user/org.couchdb.user:user2324'
 error--- authenticating for user arnabgg failed. Error: bad username/password, access denied
 warn --- the plugin method add_user in the auth plugin is deprecated and will be removed in next major release, notify to the plugin author
 error--- the user user2324 could not being added. Error: bad username/password, access denied
 http --- 409, user: null(127.0.0.1), req: 'PUT /-/user/org.couchdb.user:user2324', error: bad username/password, access denied

Here is my config.yaml file 

storage: /home/administrator/.local/share/verdaccio/storage
 
plugins: ./plugins

 
web:
  title: Verdaccio
 

 
  auth:
    htpasswd:
     file: ./htpasswd
    # Maximum amount of users allowed to register, defaults to "+inf".
    # You can set this to -1 to disable registration.
#    max_users: -1
    # Hash algorithm, possible options are: "bcrypt", "md5", "sha1", "crypt".
    # algorithm: bcrypt # by default is crypt, but is recommended use bcrypt for new installations
    # Rounds number for "bcrypt", will be ignored for other algorithms.
    # rounds: 10
    activedirectory:
     url: "ldap://10.0.100.1"
     baseDN: 'dc=sample,dc=local'
     domainSuffix: 'sample.local'

  
 
 
packages:
  '@*/*':
    # scoped packages
    access: $all
    publish: $all
    unpublish: root
#    proxy: npmjs

  'npmfolder':
 
    publish: $authenticated
    unpublish: $authenticated

 

 
server:
  keepAliveTimeout: 60
 
  
 
  security:
   api:
#     legacy: true
     jwt:
       sign:
         expiresIn: 29d
       verify:
         someProp: [value]
     web:
      sign:
        expiresIn: 1h # 1 hour by default
      verify:
         someProp: [value]

 
  userRateLimit:
   windowMs: 50000
   max: 1000

 
 
 
  

middlewares:
  audit:
    enabled: true

 
 
logs: { type: stdout, format: pretty, level: http }
 

Also please let me know if any other alternative way to add a new user

**Comment by ArnabDom** [comment, marked as answer]
@juanpicado 
please help

**Reply by juanpicado** [reply]
please help me to help you, format your question, right now is hard to read. Also as advice don't attach files because I never open them, copy the content and highlight relevant parts, describe your steps and version of what you are using.

**Reply by ArnabDom** [reply]
Thank you for your prompt reply  @juanpicado  i have updated the post
Npm and verdaccio all uptodate to the latest version
i was experimenmting with it  last few days but in the beginning, its was normal installed npm verdaccio, created user and published test package all working good but now even i forgot what do i changed exactly 
so my issue was 
i can not create new user after changed few things in the config.yaml file as i was setting max-user: -1 to disable creating new user and package access & publish to $authenticated 
and after that i can not log in again and can not create new user 

every time i try to adduser i see this on the server terminal (log)
warn --- the plugin method add_user in the auth plugin is deprecated and will be removed in next major release, notify to the plugin author
 error--- the user asdd could not being added. Error: bad username/password, access denied
http --- 409, user: null(127.0.0.1), req: 'PUT /-/user/org.couchdb.user:asdd', error: bad username/password, access denied
 http --- 127.0.0.1 requested 'GET /-/user/org.couchdb.user:asdd?write=true'
 error--- authenticating for user arnabgg failed. Error: bad username/password, access denied


and this is my .npmrc file (i have deleted the auth token just in the post so other people cant see but in the file its present)
registry=http://localhost:4873/
//localhost:4873/:_authToken=""
ca=null
strict-ssl=false
scope=@myscope
always-auth=true
@myscope:registry=http://localhost:4873/



Here is my config yaml file 

storage: /home/administrator/.local/share/verdaccio/storage
 
plugins: ./plugins
 
web:
  title: Verdaccio
   

 
  auth:
    htpasswd:
     file: ./htpasswd
     
     max_users: -1
    
    activedirectory:
     url: "ldap://10.0.100.1"
     baseDN: 'dc=sample,dc=local'
     domainSuffix: 'sample.local'
 
uplinks:
  npmjs:
    url: https://registry.npmjs.org/

 
packages:
  '@*/*':
      scoped packages
    access: $all
    publish: $all
    unpublish: root
    proxy: npmjs

  'npmfolder':
     
    access: $authenticated

    
    publish: $authenticated
    unpublish: $authenticated

     
    proxy: npmjs
 
server:
  keepAliveTimeout: 60
   
  security:
   api:
#     legacy: true
     jwt:
       sign:
         expiresIn: 29d
       verify:
         someProp: [value]
     web:
      sign:
        expiresIn: 1h # 1 hour by default
      verify:
         someProp: [value]

 
  userRateLimit:
   windowMs: 50000
   max: 1000

 

middlewares:
  audit:
    enabled: true

**Reply by juanpicado** [reply]
- which verdaccio version? 
- `activedirectory:` you are using plugins, so I guess you can figure out on your side 

The error is quite clear `bad username/password, access denied`

**Reply by ArnabDom** [reply]
Verdaccio v5.15.4
and using Ubuntu os

i was using the by-default plugin auth  and still i was having the same issue so i chaninned another plugin activedirectory:  but even when i comment out all the plugins im still having the problem 

please advice how do i fix the issue about creating new user and to be able to login 

even when i comment out  # plugins: ./plugins
and all the authorization plugins are disabled 
its still the same error

server console

 warn --- the plugin method add_user in the auth plugin is deprecated and will be removed in next major release, notify to the plugin author
 error--- the user dsaasd could not being added. Error: bad username/password, access denied
 http --- 409, user: null(127.0.0.1), req: 'PUT /-/user/org.couchdb.user:dsaasd', error: bad username/password, access denied
 http --- 127.0.0.1 requested 'GET /-/whoami'
 error--- authenticating for user arnabgg failed. Error: bad username/password, access denied
 http --- 200, user: null(127.0.0.1), req: 'GET /-/whoami', bytes: 0/2



and the other terminal where i tried to create new user

administrator@swift-sf314-42:~$ npm adduser
npm notice Log in on http://localhost:4873/
Username: dsaasd
Password: 
Email: (this IS public) test@gmail.com
npm ERR! code E409
npm ERR! 409 Conflict - PUT http://localhost:4873/-/user/org.couchdb.user:dsaasd - bad username/password, access denied

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/administrator/.npm/_logs/2022-10-13T07_50_18_192Z-debug-0.log

**Reply by juanpicado** [reply]
> please advice how do i fix the issue about creating new user and to be able to login

if you are using a plugin probably error might be at the plugin. If you are using default plugin I'd recommend remove the `rm -Rf ~/.config/verdaccio/htpasswd` file <-- I think is on that location in Ubuntu.

Or maybe you are using the wrong password, have you think on that?

**Reply by ArnabDom** [reply]
if the problem is in the plugin, do I need to disable the plugin section to resolve the issue? 
also, is it possible without any auth plugins? and if all plugins are disabled, will the default plugins still be in use in the background?

and here was another issue i forgot to mention before, i could not find htpasswd file anywhere in my system(Ubuntu 22.04.1 LTS) i checked every directory and paths but htpasswd nowhere to be visible 

 right/wrong password could be a problem but i do remember setting a very easy pass (like 1234 etc) and tried many times with different combos but nothing seems to be working

**Reply by juanpicado** [reply]
> if the problem is in the plugin, do I need to disable the plugin section to resolve the issue?

Always helps test with the defaults, because the E2E run here is with defaults, no plugins.

> also, is it possible without any auth plugins? and if all plugins are disabled, will the default plugins still be in use in the background?

if you disable plugins you get 401 by default, or bad password error.

> and here was another issue i forgot to mention before, i could not find htpasswd file anywhere in my system(Ubuntu 22.04.1 LTS) i checked every directory and paths but htpasswd nowhere to be visible

Look in your home folder it is here in hidden files.

> right/wrong password could be a problem but i do remember setting a very easy pass (like 1234 etc) and tried many times with different combos but nothing seems to be working

:) then create another usuer, maybe that helps. I'd recommend a clean installation.
